### ISSUE TITLE
node process randomly dying while running sharp

### ISSUE THREAD

**Issue by phal0r** [lovell/sharp#2473]
Are you using the latest version? Is the version currently in use as reported by `npm ls sharp` the same as the latest version as reported by `npm view sharp dist-tags.latest`?

latest stable from npm -> 0.26.3

What are the steps to reproduce?

I use sharp in the mode stream in and stream out with the pipeline *sharp().rotate().extract() *

This pipeline is cloned 10 to 20 times and then resize().webp() or resize().jpeg() is called (image type and dimensions change per clone).

Now I started a large vm (16 Intel cpus, 32 GB Ram), set UV_THREADPOOL_SIZE=128 and used sharp to process 30.000 images. I wanted to max out all the cpus and concurrently created 10 pipelines, which means about 100-200 images are processed in parallel. Memory is stable all the time, cpu load is about 80-100 and cpu usage close to 100%, which should be ok for batch processing, I guess.

Now the node process died randomly from time to time - not with the same file and not with a specific timing, just randomly. These are some of the errors:

- Bus error (core dumped)
- Error: extract_area: bad extract area
- Segmentation fault
- Premature end of JPEG file
- VipsJepg: out of order read at line

Can we a) track down, why the process is throwing those errors and b) why the node process is dying, although I attached an error handler to the stream?

If you have any hints, why this could happen, that would be awesome. If we need to play around with this setup - no problem. Just let me know, what we should try.

What is the expected behaviour?

The process should run smoothly for about 4 hours and process 30.000 files without throwing errors.

Are you able to provide a minimal, standalone code sample, without other dependencies, that demonstrates this problem?

Are you able to provide a sample image that helps explain the problem?

What is the output of running `npx envinfo --binaries --system`?

  System:
    OS: Linux 5.4 Ubuntu 20.04.1 LTS (Focal Fossa)
    CPU: (16) x64 Intel(R) Xeon(R) CPU E5-2650 v4 @ 2.20GHz
    Memory: 8.94 GB / 31.36 GB
    Container: Yes
    Shell: 5.0.17 - /bin/bash

**Comment by lovell** [comment]
Hi, if you can get a backtrace of a segfault or core dump crash via gdb that would be the most useful thing to help debug this further.

In the meantime, changes that may improve throughput include:

* Reduce `UV_THREADPOOL_SIZE` to the CPU count (16 in this case).
* Use filesystem or Buffer based I/O rather than Streams to reduce memory pressure. If you're using the same input file for multiple output files then passing filesystem paths will be much faster than `clone`.

Are the inputs all JPEG? The other errors suggest there might be some corrupt images in there.

**Comment by phal0r** [comment]
Thanks for the fast response.

> Reduce UV_THREADPOOL_SIZE to the CPU count (16 in this case).
Ok.

> Use filesystem or Buffer based I/O rather than Streams to reduce memory pressure.
I don't quite understand this. I use createReadStream from fs module to create the stream. I pipe it into sharp and after sharp it is being piped to a s3 client. If I would read all files as buffer, I would consume much more memory, no? In my observations memory consumption was stable all the time always ~30% of maximum available memory.

> If you're using the same input file for multiple output files then passing filesystem paths will be much faster than clone.
Since I would like to use fs.createReadStream, would it make sense to pipe the filestream into multiple sharp instances instead of cloning one or do you suggest to create multiple filestreams of the same file (and probably rely on filesystem caching)?

> Are the inputs all JPEG? The other errors suggest there might be some corrupt images in there.
Some are webp files. It may be the case, that some files are corrupted, but if this is the case I cannot log the filenames of the corrupted files, because the node process dies beforehand.

> if you can get a backtrace of a segfault or core dump crash via gdb that would be the most useful thing to help debug this further.
Can you point me to a tutorial or some readme on how this can be done? That would help me alot.

**Comment by lovell** [comment]
> I use createReadStream from fs module to create the stream. I pipe it into sharp and after sharp it is being piped to a s3 client.

If the input images are already on the local filesystem then providing the path to the files ensures they will only be loaded into memory by libvips at the point at which _output_ has been requested. It should also improve the ability of libvips to cache inputs.

To get a backtrace:
```
$ sudo apt-get install gdb
$ gdb node
(gdb) run /path/to/your/script.js
... wait for crash ...
(gdb) backtrace
```

**Comment by phal0r** [comment]
> If the input images are already on the local filesystem then providing the path to the files ensures they will only be loaded into memory by libvips at the point at which output has been requested. It should also improve the ability of libvips to cache inputs.

Thanks for the insights. I read in the docs for clone `Cloned instances inherit the input of their parent instance. This allows multiple output Streams and therefore multiple processing pipelines to share a single input Stream.`, so this sounded like a really good optimization performance-wise. So this means, it makes sense to use clone only on an already preprocessed input stream like in the example in contrast to a file input stream. Ok, understood.

> To get a backtrace:
Thanks, this actually is easy. I though there would be more and complex tooling involved to get this started. I will try to apply those optimizations, add gdb and report back, once I ran the workload again in the next days.

**Comment by phal0r** [comment]
I ran the script again yesterday with a subset of the data and with `UV_THREADPOOL_SIZE` set to 16 and so far no crashes. Will try the next days with more data.

Are the native threads also bound to the  process memory limit of the node process? Maybe this could be the cause of the segfaults. 

I will report again after my next round with more data.

**Comment by phal0r** [comment]
@lovell 
Any idea how to use gdb with npx? `gdb npx ts-node script.ts` yields:
```
Reading symbols from npx...
(No debugging symbols found in npx)
```
Today I encountered  a segfault again and would ike to provide debug output, but I cannot make it run with npx and ts-node

**Comment by lovell** [comment]
```
$ gdb node
(gdb) run node_modules/.bin/ts-node script.ts
```

**Comment by phal0r** [comment]
I still get segmentation faults, but less frequently and I forgot to use gdb while this happened. I will continue to execute with gdb to catch a hopefully helpful stacktrace.

**Comment by phal0r** [comment]
Also apart from the segfault, when I use this approach with streams, normal errors like `Error: extract_area: bad extract area` also lead to exiting the process completely, while I would expect it to call the error eventhandler, which it doesn't.

I also realized, that my input in the production app is not **fs.createReadStream**, but rather a readable stream provided by graphql-upload, so I cannot change it to a filepath. What would be the best approach to have this single readable stream is input and wanting multiple output streams?

Thanks for your help so far and sorry, that I mix different questions here. If you want, I can split it into different issues.

**Comment by lovell** [comment]
Uncaught Stream-related problems that exit the process are usually due to a lack of error handlers on *every* Stream that's involved.

Did you see https://sharp.pixelplumbing.com/api-constructor#clone ?

**Comment by phal0r** [comment]
I guess my setup is cluttered throughout this issues and I will clarify my current code

```
// readable stream is provided by graphql-upload library
const stream = createReadStream()

// include a duplex stream, which extracts dimensions from binary data
const sizeStream = createImageSizeStream()

const pipeline = sharp().rotate().extract({
      left: crop.x,
      top: crop.y,
      width: crop.width,
      height: crop.height,
    });

// prevent node memory leak warnings due to having more than 12 listeners
pipeline.setMaxListeners(22)

// I attach the error handler to the sharp stream, which, I expected, should catch those extraction errors
// but unfortunatel it doesn't
pipeline.on("error", (error) => {
    console.error(error);
});

// I need to render 10 or 20 renditions, so it will be cloned this often
array.forEach(() => {
     pipeline.clone().resize(size).jpeg();
     // these streams are piped to a S3 client and uploaded
})

stream.pipe(sizeStream)
sizeStream.pipe(pipeline)
```

`Error: extract_area: bad extract area` errors lead to the node process exiting, although an error handler is attached to this stream, I do not attach error handlers to the cloned streams, which are probably own stream instances. I still would expect, that the error is thrown in the first sharp stream, but maybe it bubbles to the cloned streams somehow, where the process finally crashes. I will attach error handlers there aswell and check.

**Comment by lovell** [comment]
Node.js provides `stream.pipeline()` for "forwarding errors and properly cleaning up" that is designed for exactly this kind of scenario.

https://nodejs.org/dist/latest-v15.x/docs/api/stream.html#stream_stream_pipeline_source_transforms_destination_callback

**Comment by phal0r** [comment]
Thanks for the link, I didn't know about pipelines yet. I will try to integrate them in the cloned streams. Meanwhile I also was reading about streams and stumbled upon this issue: https://stackoverflow.com/questions/19553837/node-js-piping-the-same-readable-stream-into-multiple-writable-targets

Essentially, it states, that piping one input stream into multiple streams is dangerous, when the streams consume data in different pace. I realized, that this is probably another implementation error on my side, because I have this pipeline currently:

readable input stream 
    -> imagesize stream 
          -> sharp
                -> cloned sharp stream 1
                -> cloned sharp stream 2
                -> cloned sharp stream 3
          -> s3 upload of original image

I was piping the imagesize stream into the original sharp stream AND an s3 client. This can obviously lead to internal errors due to overwriting data chunks due to different processing speeds of the s3 upload and sharp. This could potentially be the source of the segfaults imho. I will fix this and makes somes runs.

**Comment by bausano** [comment]
I also experience regular segfaults:

```
PID 7 received SIGSEGV for address: 0x5c
/usr/bin/mailmevouchers/imgur/node_modules/segfault-handler/build/Release/segfault-handler.node(+0x2cb1)[0x7fd7cc2b6cb1]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x110e0)[0x7fd7ceac70e0]
/usr/bin/mailmevouchers/imgur/node_modules/sharp/build/Release/../../vendor/8.10.0/lib/libvips.so.42(vips_image_new_from_image+0xe5)[0x7fd7c59b73e5]
/usr/bin/mailmevouchers/imgur/node_modules/sharp/build/Release/sharp.node(_ZN5sharp9OpenInputEPNS_15InputDescriptorE+0x44a)[0x7fd7cc6e2bea]
/usr/bin/mailmevouchers/imgur/node_modules/sharp/build/Release/sharp.node(_ZN14MetadataWorker7ExecuteEv+0x47)[0x7fd7cc6eb217]
/usr/bin/mailmevouchers/imgur/node_modules/sharp/build/Release/sharp.node(_ZN4Napi11AsyncWorker9OnExecuteENS_3EnvE+0x14)[0x7fd7cc6e4d24]
/usr/bin/mailmevouchers/imgur/node_modules/sharp/build/Release/sharp.node(_ZN4Napi11AsyncWorker18OnAsyncWorkExecuteEP10napi_env__Pv+0x13)[0x7fd7cc6e8673]
node[0x132ce4e]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x74a4)[0x7fd7ceabd4a4]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x3f)[0x7fd7ce7ffd0f]
Segmentation fault (core dumped)
```

**Comment by lovell** [comment]
@bausano Please note that the third-party `segfault-handler` module itself is known to cause segfaults - see #2191 - if you're still having problems, please open a new issue and provide complete steps to reproduce, perhaps as a separate repo.

**Comment by phal0r** [comment]
@lovell 
I am 99% sure, that the cause was piping twice from the same readable stream, which can lead to errors. I used the `clone-readable-stream` module to fix that and now it looks good on my site. I think, that we can close this.
